hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -0,0 +1,381 @@
+    Preconditions.checkArgument(requestFanOut >= 1);
+    if (requestFanOut > 1) {
+      masterRpcPool = Executors.newFixedThreadPool(requestFanOut,
+          new ThreadFactoryBuilder().setDaemon(true).setNameFormat("MasterRegistryRPC-%d").build());
+    }
+    parseMasterAddrs(conf);
+    rpcTimeoutNs = (int) Math.min(Integer.MAX_VALUE,
+        TimeUnit.MILLISECONDS.toNanos(conf.getLong(HBASE_RPC_TIMEOUT_KEY,
+        DEFAULT_HBASE_RPC_TIMEOUT)));
+    // TODO(HBASE-23330): Fix clients using cluster ID based token auth.
+    rpcClient = RpcClientFactory.createClient(conf, HConstants.CLUSTER_ID_DEFAULT);
+    rpcControllerFactory = RpcControllerFactory.instantiate(conf);
+  }
+
+  /**
+   * Parses the list of master addresses from the provided configuration.
+   * @param conf Configuration to parse from.
+   */
+  private void parseMasterAddrs(Configuration conf) {
+    String configuredMasters = conf.get(MASTER_ADDRS_KEY, MASTER_ADDRS_DEFAULT);
+    for (String masterAddr: configuredMasters.split(",")) {
+      masterServers.add(ServerName.valueOf(masterAddr, ServerName.NON_STARTCODE));
+    }
+    // (Pseudo) Randomized so that not all clients hot spot the same set of masters.
+    Collections.shuffle(masterServers);
+    Preconditions.checkArgument(!masterServers.isEmpty());
+  }
+
+  /**
+   * Makes a given RPC to master servers.
+   * @param rpcCall Call to make.
+   * @param debug String used for debug logging the RPC details.
+   * @param <RESP> Response type for the RPC.
+   * @param isvalidResp Used to verify if the response returned from RPC is valid.
+   * @return Optional response from the RPCs to parsed masters.
+   */
+  @VisibleForTesting
+  <RESP> Optional<RESP> doRPCs(RpcCall<RESP> rpcCall,
+      Function<RESP, Boolean> isvalidResp, String debug) {
+    if (requestFanOut == 1) {
+      // This is the most general (and default) case. We want to avoid the thread creation and
+      // synchronization overhead and hence a special optimization for this case that just loops
+      // through the available masters in that order.
+      return doSequentialRPCs(rpcCall, isvalidResp, debug);
+    }
+    return doHedgedRPCs(rpcCall, isvalidResp, debug);
+  }
+
+  /**
+   * Makes RPCs in batches of {HConstants#MASTER_REGISTRY_NUM_HEDGED_REQS_KEY}. Currently should
+   * only be used if the request fan out is > 1. For the default case, refer to doSequentialRPCs().
+   */
+  private <RESP> Optional<RESP> doHedgedRPCs(RpcCall<RESP> rpcCall,
+      Function<RESP, Boolean> isvalidResp, String debug) {
+    Preconditions.checkState(requestFanOut > 1);
+    Preconditions.checkNotNull(masterRpcPool);
+    int i = 0;
+    while (i < masterServers.size()){
+      // Each iteration of loop picks requestFanOut masters
+      int subListSize = Math.min(masterServers.size(), i + requestFanOut);
+      List<ServerName> masterSubList = masterServers.subList(i, subListSize);
```

Comment:
Shouldn't the list be randomized again? Or we're hitting the sublists deterministically. Make a private list at top of function and shuffle?